### PR TITLE
Use ErrorProne to validate our code

### PR DESCRIPTION
**Pull request by lptr**
This introduced Error Prone to our build. Some considerations:

- We are checking the `main` source set only by default. In some subprojects we have Java code in our test code, but joint Groovy compilation is not happy when we both enable experimetnal incremental compilation _and_ add the annotation processor for Error Prone.
- Caveat: Kotlin and Groovy code is not checked by Error Prone, only Java.

We chose to globally ignore some warnings, at least for the time being. We should discuss if more/fewer checks should be ignored globally.

- [MissingSummary](https://errorprone.info/bugpattern/MissingSummary) -- we have other checks to ensure Javadoc is present on our public APIs
- [InjectOnConstructorOfAbstractClass](https://errorprone.info/bugpattern/InjectOnConstructorOfAbstractClass) and [JavaxInjectOnAbstractMethod](https://errorprone.info/bugpattern/JavaxInjectOnAbstractMethod) -- we actually use "inject to abstract" as a pattern in Gradle
- [JavaUtilDate](https://errorprone.info/bugpattern/JavaUtilDate) -- we are fine using `java.util.Date`.
- [JdkObsolete](https://errorprone.info/bugpattern/JdkObsolete) -- this is a nice check that flags old data types like `Vector` and `Hashtable`, but it also flags `LinkedList` in a lot of places. We don't want to blindly replace the latter with `ArrayList`, without an actual detected performance bottleneck.
- [EqualsGetClass](https://errorprone.info/bugpattern/EqualsGetClass) -- this is a question of preference, and most of Gradle's code is written not matching that preference.

## Process of resolving warnings

Initially there was upwards of 1700 warnings. We've fixed all the missing `@Override` warnings, and added project-level exclusions for most of the remaining warnings, like so:

```kotlin
errorprone {
    disabledChecks.addAll(
        "ReferenceEquality", // 3 occurrences
        "UnnecessaryParentheses", // 1 occurrences
        "UnusedMethod", // 1 occurrences
    )
}
```

The idea is that code owners (or folks just passing by) could take some of the disabled warnings, remove the disabling, and fix the warnings. If we keep doing this, eventually we'll have no warnings left, and profit!

### CODE REVIEW

**Comment by JLLeitschuh** [comment]
Curious about the choice to suppress for an entire project, instead of using annotations to suppress on specific scopes.

Also, curious if OpenRewrite/Moderne can help you cleanup outstanding alerts

File: build-logic-commons/gradle-plugin/src/main/kotlin/gradlebuild.code-quality.gradle.kts [hunk 1]
```diff
@@ -21,6 +24,61 @@ plugins {
...
+        )
+
+        project.tasks.named<JavaCompile>(this.compileJavaTaskName) {
+            options.errorprone {
+                isEnabled = extension.enabled
+                checks.set(errorproneExtension.disabledChecks.map {
+                    it.stream().collect(Collectors.toMap(
+                        { check -> check },
+                        { _ -> CheckSeverity.OFF }
+                    ))
```
Comment (vlsi) [inline build-logic-commons/gradle-plugin/src/main/kotlin/gradlebuild.code-quality.gradle.kts:70]: WDYT of the following?

```suggestion
                    it.associateWith { CheckSeverity.OFF }
```

**Comment by vlsi** [comment]
@JLLeitschuh, OpenRewrite would be cool, and it would be great to automatically add `@Override` annotations instead of just complaining "oh, you miss `@Override`".

Unfortunately, the current OpenRewrite's Gradle plugin takes significant time to execute:
1) It [assumes to be applied to root project](https://github.com/openrewrite/rewrite-gradle-plugin/issues/183#issuecomment-1668656797) and it is hard to "process a single subproject only"
1) It spends [~30 seconds to initialize when scanning the jars](https://github.com/openrewrite/rewrite/issues/3899)
1) It misses [incremental processing](https://github.com/openrewrite/rewrite-gradle-plugin/issues/256). In other words, even repeated executions take time to process the files again. I want repeated executions to be `UP-TO-DATE` or `FROM-CACHE`.

I'm working on integrating OpenRewrite for Apache JMeter, and based on the above issues I am leaning towards reimplementing rewrite-gradle-plugin, see https://github.com/apache/jmeter/pull/6217

File: build-logic-commons/gradle-plugin/src/main/kotlin/gradlebuild.code-quality.gradle.kts (hunk 1 above)
Comment (lptr) [reply build-logic-commons/gradle-plugin/src/main/kotlin/gradlebuild.code-quality.gradle.kts:70]: Looks good, thanks.

**Comment by lptr** [comment]
> Curious about the choice to suppress for an entire project, instead of using annotations to suppress on specific scopes.

The goal is to start using Error Prone as much as we can already to spot new regressions. Doing this allowed us to do it without a lot of noise, and a quick overview of how bad a subproject is with current warnings. This way it is also clear when a subproject is rid of all the initial warnings.

> Also, curious if OpenRewrite/Moderne can help you cleanup outstanding alerts

I'm not sure about that. The trivial ones, maybe, though stuff like missing `@Override` are simple enough to fix anyway. There isn't a single type of warning where we would need to repeat the same kind of change many times that would make investing into OpenRewrite a rational choice I think. I expect the work from now is going to be mostly to do one-off cleanups, and long discussions in the team about whether checks like using `getClass()` in `equals()` make sense for us.

**Comment by lptr** [comment]
@bot-gradle test and merge please.

**Comment by JLLeitschuh** [comment]
> I'm working on integrating OpenRewrite for Apache JMeter, and based on the above issues I am leaning towards reimplementing rewrite-gradle-plugin, see apache/jmeter#6217

Facinating! Why not upstream your improvements into the official OpenRewrite plugin?

Also, CC: @jkshneider, @walkingrufus I think this might interest you both

**Comment by lptr** [comment]
@bot-gradle merge this please.

**Comment by wakingrufus** [comment]
> > I'm working on integrating OpenRewrite for Apache JMeter, and based on the above issues I am leaning towards reimplementing rewrite-gradle-plugin, see [apache/jmeter#6217](https://github.com/apache/jmeter/pull/6217)
> 
> Facinating! Why not upstream your improvements into the official OpenRewrite plugin?
> 
> Also, CC: @jkshneider, @walkingrufus I think this might interest you both
This is getting off-topic, but I have done some proof of concepts around how to make a rewrite plugin using the worker API and properly wiring inputs/outputs, which would work for parallel and multi-project builds. it is possible, but I have not yet put it together in a production-ready way. Although my implementation involved copy and pasting a decent amount of code from the existing OSS plugin, i don't know if evolving that plugin would be the best way to go. I think it would it would be easier if the project were started from scratch. I am happy to discuss more in the Open-Rewrite slack.

**Comment by reinsch82** [comment]
@bot-gradle merge

**Pull request merged**